include/seqan3/search/search.hpp

```diff
@@ -89,6 +93,40 @@ struct search_configurator
 
         return cfg2;
     }
+
+    /*!\brief Chooses the appropriate search algorithm depending on the index.
+     * \tparam configuration_t The type of the search configuration.
+     * \tparam index_t The type of the index.
+     * \param[in] cfg The search configuration object that is passed to the algorithm.
+     * \param[in] index The index that is passed to the algorithm.
+     * \returns A search algorithm.
+     *
+     * \details
+     *
+     * If the `index_t` models seqan3::bi_fm_index_specialisation, then the
+     * seqan3::detail::search_scheme_algorithm is chosen. Otherwise, the
+     * detail::unidirectional_search_algorithm is chosen.
+     */
+    template <typename configuration_t, typename index_t>
+    static auto configure_algorithm(configuration_t const & cfg, index_t const & index)
+    {
+        if constexpr (bi_fm_index_specialisation<index_t>)
+        {
+            using algorithm_t = search_scheme_algorithm<configuration_t,
+                                                        index_t,
+                                                        policy_max_error,
+                                                        policy_result_builder>;
```

Comment:
This shows that you didn't really use policies as intended.

Both classes use the same policies.

I'm not sure how all searches are, but from what I could see they all have an underlying search, like trivial search and bidirectional search (which could be policies) and you would get a single search_algorithm, that would look like this:

```c++
// pseudo code:
using algorithm_t = index_based_search_algorithm<
    configuration_t,
    index_t,
    // provides search_impl (currently search_trivial, search_algo_bi)
    is_bi_fm_index ? policy_search_scheme :
                     policy_unidirectional,
    // provides operator()
    switch(search_mode):
        mode == all : policy_search_mode_all
        mode == best_hits : policy_search_mode_best_hits
        mode == all_best_hits : policy_search_mode_all_best_hits
        mode == strata_hits : policy_search_mode_strata_hits,
    policy_max_error_state,
    policy_make_results>;
```

Reply:
When I read through the alignment algorithm and its policies and asked Rene about the design decisions, I thought to understand that policies are **reusable or customisable parts** of an algorithm. From this definition, I don't think search_trivial/unidirectional and search_algo_bi/search_scheme_algorithm should be designed as policies. They are the algorithms itself, no?
Additionally, I think that in the future, we might have actual differences between the search modi of the search algorithms uni/bi. So we would need to refactor this back into algorithms.

These are just my thoughts on this. Either way, could this be part of a later restructuring?

Reply:
A different behaviour for a mode could be implemented by a different underlying search via a policy which will be used conditionally. I find it funny, because from what I have seen, the underlying search is the **reusable or customisable part** of the algorithm. This might be a "Wo ein Wille ist, ist auch ein Weg" issue :)